### PR TITLE
Alexmont/dpe widget

### PR DESCRIPTION
**Pull request by alexmontAmazon**
- implemented expander state save/restore
- implemented additional RPE expander attributes
- small fixes and const correctness

### CODE REVIEW

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/PropertyEditorNodes.cpp [hunk 1]
```diff
@@ -42,8 +42,14 @@ namespace AZ::DocumentPropertyEditor::Nodes
         system->RegisterNodeAttribute<NodeWithVisiblityControl>(NodeWithVisiblityControl::Visibility);
 
         system->RegisterNode<Adapter, NodeWithVisiblityControl>();
+
         system->RegisterNode<Row, NodeWithVisiblityControl>();
 
+        for (const auto& rowAttribute : Row::RowAttributes)
+        {
+            system->RegisterNodeAttribute<Row>(rowAttribute);
```
Comment (NicholasVanSickle) [inline Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/PropertyEditorNodes.cpp:50]: Cool, we might even be able to generalize this to an optional `Attributes` field for node definitions to make registration easier. (Not for this PR)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 2]
```diff
@@ -347,6 +353,37 @@ namespace AzToolsFramework
     {
         Clear();
 
+        // determine whether this node should be expanded
+        auto forceExpandAttribute = domArray.FindMember(AZ::Dpe::Nodes::Row::ForceAutoExpand.GetName());
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:357]: Could also do: `AZ::Dpe::Nodes::Row::ForceAutoExpand.ExtractValueFromNode(domArray).value_or(false)` (same below with `AutoExpand`)

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 3]
```diff
@@ -390,6 +427,13 @@ namespace AzToolsFramework
                 domOperation.GetType() == AZ::Dom::PatchOperation::Type::Replace)
             {
                 const auto childIterator = m_domOrderedChildren.begin() + childIndex;
+                DPERowWidget* rowToRemove = qobject_cast<DPERowWidget*>(*childIterator);
+                if (rowToRemove)
+                {
+                    // we're removing a row, remove any associated saved expander state
+                    GetDPE()->RemoveExpanderStateForRow(this);
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:434]: Shouldn't this be `rowToRemove`?

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 4]
```diff
@@ -583,31 +631,106 @@ namespace AzToolsFramework
...
 
-    AZ::Dom::Value DocumentPropertyEditor::GetDomValueForRow(DPERowWidget* row) const
+    void DocumentPropertyEditor::SetSavedExpanderStateForRow(DPERowWidget* row, DocumentPropertyEditor::ExpanderState expanderState)
     {
-        AZStd::vector<size_t> reversePath;
-        const DPERowWidget* thisRow = row;
-        const DPERowWidget* parentRow = thisRow->m_parentRow;
+        // Get the index of each dom child going up the chain. We can then reverse this
+        // and use these indices to mark a path to expanded/collapsed nodes
+        AZStd::vector<size_t> reversePath = GetPathToRoot(row);
```
Comment (NicholasVanSickle) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:638]: I'm good with this as-is, but I think `DomPrefixTree` can definitely make your life easier if you ever need to do something like this, it handles the "look up the nearest ancestor that explicitly has a value" behavior here.

**Review by AMZN-daimini — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 2 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp]: fixed! Make sure it's to your liking

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 3 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp]: good catch! Fixed.

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp (hunk 4 above)
Comment (alexmontAmazon) [reply Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:638]: noted, thanks

File: Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/PropertyEditorNodes.cpp (hunk 1 above)
Comment (alexmontAmazon) [reply Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/PropertyEditorNodes.cpp:50]: sure, sounds good.

**Review by NicholasVanSickle — APPROVED** [review]
Looks good to me!

File: Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp [hunk 5]
```diff
@@ -8,13 +8,15 @@
 #include "DocumentPropertyEditor.h"
 
 #include <AzQtComponents/Components/Widgets/ElidingLabel.h>
+#include <QCheckBox>
+#include <QDebug>
```
Comment (cgalvan) [inline Code/Framework/AzToolsFramework/AzToolsFramework/UI/DocumentPropertyEditor/DocumentPropertyEditor.cpp:12]: Can we remove this `QDebug` ?

**Review by cgalvan — APPROVED** [review]

**Pull request merged**